### DISCUSSION TITLE
PreCommit should fix mixed-line-endings automatically / Add .gitattribute

### DISCUSSION THREAD

**Discussion by CarliJoy** [pyscaffold/pyscaffold#337] [Ideas]
## Description of your problem

Copy stuff from somewhere to PyCharm from a source with different line ending.

This will prevent me from committing at all.
Users with not so much experience (not everybody who writes python packages is a computer geek) will have a hard time getting the commit to work.
It would be nice if the setting is more forgiving, so there is an error but an auto fix.

**Please provide any additional information below.**
Suggestion: Change the `.pre-commit-config.yaml` to
```
  - id: mixed-line-ending
    args: ['--fix=auto']
```

I don't know if that catches all edge cases especially working with autoclrf.
Maybe it would be nice to introduce a .gitattribute instead?
See also: https://stackoverflow.com/questions/2825428/why-should-i-use-core-autocrlf-true-in-git

It probably depends also a lot on the user settings but maybe we still could safely add some files (like *.sh) to a gitattribute file?

PS: Github offers Discussions now. This issue would fit right into them.

**Comment by abravalheri** [comment]
Thanks @CarliJoy that is indeed a nice idea.

Do you think that would be any drawbacks if we proceed with the suggested implementation?

**Reply by CarliJoy** [reply]
Actually I am **not** experienced enough to answer this questions. 
I personally never had problems with the setting in opposite to the pyscaffold default.
Given that you thinking of an interactive mode also for inexperienced users I guess auto is the only option ;-)

pre-commit has auto set as default setting:
https://github.com/pre-commit/pre-commit-hooks#mixed-line-ending
so they seem confident about it.

If this trouble deciding we could ask in stackoverflow about experiences.

**Reply by FlorianWilhelm** [reply]
I would also second the idea of including this feature. Not sure how pre-commit determines the right file endings but just the heuristic of converting line endings to the majority seems like an easy and almost always working approach. 

In almost all default configurations in PyScaffold we have some nice comments and hints to help users. We could also do it here like:
```yaml
  - id: mixed-line-ending
    args: ['--fix=auto']  # replace 'auto' with 'lf' to enforce Linux/Mac line endings or 'crlf' for Windows
```

**Reply by abravalheri** [reply]
Thank you very much guys! That definetely seems to be the way forward then. I also like the comment suggested by @FlorianWilhelm since it gives to the users a way of solving any eventual problems that might appear.

**Comment by julie777** [comment]
I think that all text files should be checked in with Unix standard "\n" line endings. This should be forced.

### Implementing The Requirement

A good way to ensure the requirement is met, and to make sure it doesn't depend on precommit or local developer settings it to add the following two files to the project.

> With this config-based approach, you can ensure that your line endings remain consistent in your codebase regardless of what operating systems or local Git settings your developers use since this file takes priority. 

.gitattributes
```
* text=auto
```

> Many code editors such as vscode will use this file if present to keep workspace file line endings correct.

.editor.config
```
root = true

[*]
end_of_line = lf
```

(If we decide to do this, I will gladly add the new files and make sure all existing files in the repo are normalized.)

### Reasoning

Here is a good [reference](https://www.aleksandrhovhannisyan.com/blog/crlf-vs-lf-normalizing-line-endings-in-git/)

I have never had a problem with always using "\n" for text files, in the context of development. Every tool I have used works just fine. That said, git is configurable to switch files in the working directory to platform specific line endings if someone wants to work that way locally.

As a project we should require proper LF in the repo because that is the most "standard". We only have to add .gitattributes to the project to enforce it. It doesn't require a precommit hook that way.

Git diff also doesn't show differences if working directory files have different line endings from committed files.

### Background

Originally Unix used LF. Then dos came along and used CRLF. Of course Apple had to be different and so they used CR for MacOS. When MacOS switched to having a Unix-like core, Apple switched to LF. So only Windows is the oddball now.
